From: Gerald Schaefer <gerald.schaefer@de.ibm.com>
Subject: s390/pci: set error state for unavailable functions
Patch-mainline: v3.14-rc1
Git-commit: 0c0c2776926018e7560b99e921467aea1115d03b
References: fate#315311, LTC#92765

Summary:     s390: add pci support
Description: Add pci support for System z.

Upstream-Description:

             s390/pci: set error state for unavailable functions

             If we receive a notification that a pci function became unavailable we clean
             up by removing the pci device. This can confuse the driver since the function
             is already unaccessible. Improve this situation by setting an appropriate
             error_state.

             Reviewed-by: Gerald Schaefer <gerald.schaefer@de.ibm.com>
             Signed-off-by: Sebastian Ott <sebott@linux.vnet.ibm.com>


Signed-off-by: Gerald Schaefer <gerald.schaefer@de.ibm.com>
Acked-by: John Jolly <jjolly@suse.de>
---
 arch/s390/pci/pci_event.c |    6 +++++-
 1 file changed, 5 insertions(+), 1 deletion(-)

--- a/arch/s390/pci/pci_event.c
+++ b/arch/s390/pci/pci_event.c
@@ -102,8 +102,12 @@ static void __zpci_event_availability(st
 
 		break;
 	case 0x0304: /* Configured -> Standby */
-		if (pdev)
+		if (pdev) {
+			/* Give the driver a hint that the function is
+			 * already unusable. */
+			pdev->error_state = pci_channel_io_perm_failure;
 			pci_stop_and_remove_bus_device(pdev);
+		}
 
 		zpci_disable_device(zdev);
 		zdev->state = ZPCI_FN_STATE_STANDBY;
